### PR TITLE
[IRGen] Use a more precise computation for the kind of reference count.

### PR DESCRIPTION
**Pull request by DougGregor**
TypeBase::usesNativeReferenceCounting() was doing a lot of work to
find the class that a type refers to, then determine whether it
would use the native reference-counting scheme. Its primary caller
in IRGen would use an overly-conservative approximation to decide
between the “Objective-C” and “unknown” cases, which resulted in
uses of “unknown” reference counting for some obviously-ObjC cases
(e.g., values of “NSObject”).

Moreover, the approximation would try to call into the type checker
(because it relied unnecessarily on the superclass *type* of a class
declaration), causing an assertion.

Fixes rdar://problem/42828798.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please test and merge

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: lib/AST/Type.cpp [hunk 1]
```diff
@@ -4098,17 +4112,17 @@ bool TypeBase::usesNativeReferenceCounting(ResilienceExpansion resilience) {
     assert(archetype->requiresClass() ||
            (layout && layout->isRefCounted()));
     if (auto supertype = archetype->getSuperclass())
-      return supertype->usesNativeReferenceCounting(resilience);
-    return ::doesOpaqueClassUseNativeReferenceCounting(type->getASTContext());
+      return supertype->getReferenceCounting(resilience);
+    return getUnknownObjectReferenceCounting();
```
Comment (slavapestov) [inline lib/AST/Type.cpp]: In theory we could have a layout constraint but I guess those aren't fully plumbed through yet anyway...

File: lib/AST/Type.cpp [hunk 2]
```diff
@@ -4145,6 +4159,10 @@ bool TypeBase::usesNativeReferenceCounting(ResilienceExpansion resilience) {
   llvm_unreachable("Unhandled type kind!");
 }
 
+bool TypeBase::usesNativeReferenceCounting(ResilienceExpansion resilience) {
```
Comment (slavapestov) [inline lib/AST/Type.cpp]: Is this used in enough places to justify its existence?

File: lib/IRGen/GenClass.cpp [hunk 3]
```diff
@@ -63,22 +63,7 @@ using namespace irgen;
 /// What reference counting mechanism does a class-like type have?
 ReferenceCounting irgen::getReferenceCountingForType(IRGenModule &IGM,
```
Comment (slavapestov) [inline lib/IRGen/GenClass.cpp]: I would just fold this function into callers

File: lib/AST/Type.cpp [hunk 4]
```diff
@@ -4046,21 +4047,31 @@ bool UnownedStorageType::isLoadable(ResilienceExpansion resilience) const {
...
-  return !theClass->hasClangNode();
+
+  return theClass->hasClangNode()
+           ? ReferenceCounting::ObjC
+           : ReferenceCounting::Native;
 }
 
-bool TypeBase::usesNativeReferenceCounting(ResilienceExpansion resilience) {
+ReferenceCounting TypeBase::getReferenceCounting(
+                                              ResilienceExpansion resilience) {
```
Comment (slavapestov) [inline lib/AST/Type.cpp]: I don't think @jrose-apple wants to allow changing the superclass of a class from Objective-C to non-Objective-C... wouldn't it mess up too many things? Might want to just drop the resilience expansion parameter here

**Comment by slavapestov** [comment]
Thanks for fixing my "NSObject is unknown refcounted" snafu!

File: lib/AST/Type.cpp [hunk 5]
```diff
@@ -4065,12 +4065,10 @@ ReferenceCounting TypeBase::getReferenceCounting(
...
-  // Determine which reference-counting scheme to use for an unknown object.
-  bool objCInterop = ctx.LangOpts.EnableObjCInterop;
-  auto getUnknownObjectReferenceCounting = [objCInterop] {
-    return objCInterop ? ReferenceCounting::Unknown
-                       : ReferenceCounting::Native;
-  };
+  // In the absence of Objective-C interoperability, everything uses native
+  // reference counting.
+  if (!ctx.LangOpts.EnableObjCInterop)
+    return ReferenceCounting::Native;
```
Comment (slavapestov) [inline lib/AST/Type.cpp]: No bridge objects?

File: lib/AST/Type.cpp (hunk 4 above)
Comment (jrose-apple) [reply lib/AST/Type.cpp]: Arguably, "insert superclasses in the hierarchy" could include "insert a new base class", but I don't think that's a *good* idea or anything.

File: lib/AST/Type.cpp (hunk 2 above)
Comment (DougGregor) [reply lib/AST/Type.cpp]: No, it isn't. I'll fold it into callers.

File: lib/AST/Type.cpp (hunk 4 above)
Comment (DougGregor) [reply lib/AST/Type.cpp]: I'll kill off this parameter. We're not using it now, and it doesn't seem that we're likely to do so.

File: lib/AST/Type.cpp (hunk 5 above)
Comment (DougGregor) [reply lib/AST/Type.cpp]: When we're not interoperating with Objective-C, the bridged classes are also defined in Swift, so it's still native everywhere.

File: lib/IRGen/GenClass.cpp (hunk 3 above)
Comment (DougGregor) [reply lib/IRGen/GenClass.cpp]: Yeah. PR coming shortly addressing all of the feedback here. Thanks!

**Pull request merged**